app/controllers/samples_controller.rb

```diff
@@ -1,6 +1,7 @@
 class SamplesController < ApplicationController
```

Comment:
This logic is getting significant. Maybe not right now, but it would be great to see some more tests.

Reply:
agree. i have a few more things i want to add in for rerunning pipelines, etc. i will add all the test then.